Client-side violations for money request updates

lindboe opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details

Adds optimistic client-side updates for money requests for missing tags, categories, or disabled tags/categories. This PR handles edits, while https://github.com/Expensify/App/pull/32528 handled creation of money requests. Requires setting up a policy in old dot and configuring it to have a workspace (see testing instructions). This PR does not make any visible changes; that is scoped to other tickets.

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/31093
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

#### Setup:

1. Follow the instructions [here](https://expensify.slack.com/archives/C05EQBQG6E5/p1695182901255769) to set up a policy with a workspace in New Dot.
1. Within the policy in Old Dot, navigate Settings > Categories and verify that People must categorize expenses is enabled, and make sure there's at least two categories on the policy
1. Within the policy in Old Dot, navigate Settings > Tags and verify that People must tag expenses is enabled, and make sure there's at least two tags enabled on the policy

#### Instructions for all tests

**Expectations for violations, generally:**

1. Violations should be visible when viewing the request on each menu item with text and a RBR indicator. Here, we're generally concerned the following violations that we calculate client-side: missing category, missing tag, category out of policy, and tag out of policy. Other visible violations should not be affected by this PR.

**Testing the violations were stored correctly (web only):**

1. Navigate to the `ONYXDB` -> `keyvaluepairs` within your browser storage tab
1. Identify the transaction that was created with the dollar amount you specified, should look like `transactions_someID`. If you open chrome dev tools before making the money request, you can inspect the response which should have the transactionID
1. Verify there's a transaction violations object with the corresponding ID like `transactionViolations_SameIDasTransaction`, with the array content specified by the test

#### Test no client-side violations

1. Create a request and request money from the policy you created in setup. Set an amount, merchant, category, and tag. Ensure your request amount is not above the maximum set by the policy, and not above the amount the policy requires for a receipt.
3. (Web) Once the request is created, check your browser storage tab. You should see either no matching `transactionViolations_ID` key, or, if you've navigated to edit the request, you may see a transaction violations list with only a `cashExpenseWithNoReceipt` violation.
4. Edit the request's date and save. There should be no visible errors on the request or the LHN.
5. (Web) inspect the transaction violations key again. The violations list should still only contain a `cashExpenseWithNoReceipt` violation.

#### Test missing tag

1. Create a request and request money from the policy you created in setup. Set an amount, merchant, and category, but DO NOT set tag. Save the request.
2. (Web) inspect the transaction violations object. You should see a matching `transactionViolations_ID` key, with an array containing a `missingTag` violation, and possibly also a `cashExpenseWithNoReceipt` violation.
3. Edit the request to add a tag. There should be no visible errors on the request or the LHN.
4. 2. (Web) inspect the transaction violations key again. You should see either no matching `transactionViolations_ID` key, or, if you've navigated to edit the request, you may see a transaction violations list with only a `cashExpenseWithNoReceipt` violation.

#### Test missing category

1. Create a request and request money from the policy you created in setup. Set an amount, merchant, and tag, but DO NOT set category. Save the request.
2. (Web) inspect the transaction violations object. You should see a matching `transactionViolations_ID` key, with an array containing a `missingCategory` violation, and possibly also a `cashExpenseWithNoReceipt` violation.
3. Edit the request to add a category. There should be no visible errors on the request or the LHN.
4. 2. (Web) inspect the transaction violations key again. You should see either no matching `transactionViolations_ID` key, or, if you've navigated to edit the request, you may see a transaction violations list with only a `cashExpenseWithNoReceipt` violation, or no violations.

#### Test tag out of policy

1. Create a request and request money from the policy you created in setup. Set an amount, merchant, and tag and category that are enabled on old dot. Save the request.
2. On old dot, disable the tag you used. Refresh new dot, and ensure a `tagOutOfPolicy` violation is shown in the transaction violations list.
3. Edit the request to use a tag that is enabled in the policy, and ensure that `tagOutOfPolicy` is no longer in the transaction violations list.

#### Test category out of policy

1. Create a request and request money from the policy you created in setup. Set an amount, merchant, and tag and category that are enabled on old dot. Save the request.
2. On old dot, disable the category you used. Refresh new dot, and ensure a `categoryOutOfPolicy` violation is shown in the transaction violations list.
3. Edit the request to use a category that is enabled in the policy, and ensure that `categoryOutOfPolicy` is no longer in the transaction violations list.

### Offline tests

Repeat the tests outlined above while completely offline, except for "tag/category out of policy", because you can't change policy while offline. Different steps for those tests are provided below.

Then, in addition, after each test, go online and verify the transactionViolations key remains the same, with the potential addition of a `cashExpenseWithNoReceipt` violation, or no violations.

#### Test tag out of policy (offline)

1. Create a request and request money from the policy you created in setup. Set an amount, merchant, and tag and category that are enabled on old dot. Save the request.
2. On old dot, disable the tag you used. Refresh new dot, and ensure a `tagOutOfPolicy` violation is shown in the transaction violations list.
3. On new dot, go offline. Select a tag that has remained enabled on your policy on the request instead. Save the request, and ensure the transaction violations list no longer has a `tagOutOfPolicy` violation.
4. Go online, and ensure that the transaction violations list does not change.

#### Test category out of policy (offline)

1. Create a request and request money from the policy you created in setup. Set an amount, merchant, and tag and category that are enabled on old dot. Save the request.
2. On old dot, disable the category you used. Refresh new dot, and ensure a `categoryOutOfPolicy` violation is shown in the transaction violations list.
3. On new dot, go offline. Select a category that has remained enabled on your policy on the request instead. Save the request, and ensure the transaction violations list no longer has a `categoryOutOfPolicy` violation.
4. Go online, and ensure that the transaction violations list does not change.

#### Test failure to edit request (web only, requires Chrome)

1. Create a request and request money from the policy you created in setup. Set an amount, merchant, and category, and tag. Save the request.
2. Go to old dot and disable the category you used.
3. Refresh new dot.
4. Inspect the corresponding `transactionViolations_` key to ensure there is a `categoryOutOfPolicy` violation there.
6. Now, cause the request to the backend to fail:
7. With the Network tab open in Chrome devtools, edit a DIFFERENT request's category. Look for a request entry matching `api?command=EditMoneyRequest` (it is possible that instead it might be a more specific command in the future, like `UpdateMoneyRequestTag`). Right-click and select "Block request URL". This will cause subsequent calls to this URL to fail.
7. Go offline and edit the request to use an enabled category. Save the request, and inspect the transaction violations key. The `categoryOutOfPolicy` violation should be gone.
8. Go online and allow the request to be sent and fail.
9. Wait a little bit for the request to fail after several retries (you should see an error in the transaction thread), and verify that the transaction violations list in Onyx is reset to what it was before you fixed the request (i.e., it contains a `categoryOutOfPolicy` violation).
10. Unblock the endpoint you blocked before, and again update the request to use an enabled category. The transactionViolations list should be updated to remove the `categoryOutOfPolicy` violation.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
11. Upload an image via copy paste
12. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

Same as tests

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<img width="508" alt="Screenshot 2024-01-16 at 3 44 55 PM" src="https://github.com/Expensify/App/assets/5252268/806e245a-0b1c-4c1d-abf5-fad9ac287e21">

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<img width="507" alt="Screenshot 2024-01-16 at 3 48 29 PM" src="https://github.com/Expensify/App/assets/5252268/057b8b5a-48db-46dc-8597-5f30f84df633">

</details>

<details>
<summary>iOS: Native</summary>

<img width="409" alt="Screenshot 2024-01-16 at 3 53 16 PM" src="https://github.com/Expensify/App/assets/5252268/19a35ccd-8806-4f61-921d-435be0aed3fd">

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<img width="413" alt="Screenshot 2024-01-16 at 4 04 48 PM" src="https://github.com/Expensify/App/assets/5252268/528898ee-3eae-4a48-adf2-db7c702e14a2">


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

![Test_1](https://github.com/Expensify/App/assets/5252268/de30b45f-9c75-432c-ad04-2cc1af3e23e9)

![Test2_beforeEdit](https://github.com/Expensify/App/assets/5252268/2bf8c300-d734-4a33-96bf-b66133a3bc8a)
![Test2_afterEdit](https://github.com/Expensify/App/assets/5252268/f4f4eab2-2257-48c2-afc5-9f793afb9fa7)

![Test3_beforeEdit](https://github.com/Expensify/App/assets/5252268/7bb15576-5775-4796-9faf-cbaa6d2742bf)
![Test3_afterEdit](https://github.com/Expensify/App/assets/5252268/bb43b5f8-92fc-4657-8c13-54acc266fae6)

![Test4_beforeEdit](https://github.com/Expensify/App/assets/5252268/e1ca639d-e1a6-4d0f-9bce-f011f686970a)
![Test4_afterEdit](https://github.com/Expensify/App/assets/5252268/743bc8eb-c313-4e57-81a1-05c7ec539c47)

**Tag out of policy tests** 
<img width="1283" alt="TagOutOfPolicy_SetupTags" src="https://github.com/Expensify/App/assets/5252268/adb6787b-3710-4d14-9450-238ed0ce8e61">
<img width="1281" alt="TagOutOfPolicy_BeforeEdit" src="https://github.com/Expensify/App/assets/5252268/62c0c05c-6dae-4d98-82a2-1bd3a3405281">
<img width="1282" alt="TagOutOfPolicy_AfterEditOffline" src="https://github.com/Expensify/App/assets/5252268/e070efe4-d7fd-4f02-a6ad-f3e3ccf33939">
<img width="1279" alt="TagOutOfPolicy_AfterEditOnline" src="https://github.com/Expensify/App/assets/5252268/dcab99c5-89c5-4eeb-85de-86e5355b1ad5">

**Category out of policy tests** 
<img width="1282" alt="CategoryOutOfPolicy_BeforeEdit" src="https://github.com/Expensify/App/assets/5252268/93e4ac8e-fa5c-4664-a3bf-eee21d7bb3f4">
<img width="1282" alt="CategoryOutOfPolicy_AfterEditOffline" src="https://github.com/Expensify/App/assets/5252268/29f90fe0-7731-43f5-8ea8-940ed447ef87">
<img width="1281" alt="categoryOutOfPolicy_AfterEditOnline" src="https://github.com/Expensify/App/assets/5252268/02fc286f-330d-4af2-8a85-851e158edd99">

**Failure data tests**

<img width="1279" alt="FailureData_Before" src="https://github.com/Expensify/App/assets/5252268/ab0c6277-9ce9-4254-b70c-9fda7be0dcd8">
<img width="1279" alt="FailureData_Optimistic" src="https://github.com/Expensify/App/assets/5252268/d850f0bb-54ed-4bfd-add4-02c37a2a6455">
<img width="1281" alt="FailureData_Revert" src="https://github.com/Expensify/App/assets/5252268/393724a2-1a8f-4e79-b608-a7cd5b9232e5">


</details>

<details>
<summary>MacOS: Desktop</summary>

<img width="1191" alt="Screenshot 2024-01-16 at 3 51 22 PM" src="https://github.com/Expensify/App/assets/5252268/88170173-eef5-4485-b170-d8d50826ae97">
<img width="1188" alt="Screenshot 2024-01-16 at 3 51 37 PM" src="https://github.com/Expensify/App/assets/5252268/d7757055-aa09-4987-8997-34b834c9cef3">


</details>

lindboe committed:
Add optimistic violations to money request edits

Update everywhere that accesses an arg on policy to use {} as default arg for safety, since we can't use optional chaining

lindboe committed:
Merge remote-tracking branch 'upstream/main' into lindboe/violations/money-request-updates

lindboe committed:
Lint fix

lindboe committed:
Fix bugs

lindboe committed:
Fix data passing

lindboe committed:
Temp fix for tag issue

lindboe committed:
Use full policyTags data as ViolationsUtils input

lindboe committed:
getViolationsOnyxData requires new transaction data but previous transaction violations. Fix and make that more clear with argument names

lindboe committed:
Improve types

lindboe committed:
Fix case of passing empty object to policyTags

lindboe committed:
We should only run getViolationsOnyxData if there's a policy

lindboe commented on src/libs/ViolationsUtils.ts:
```diff
@@ -11,10 +11,10 @@ const ViolationsUtils = {
         transactionViolations: TransactionViolation[],
         policyRequiresTags: boolean,
-        policyTags: PolicyTags,
+        policyTagList: MaybePolicyTagList,
```
@cead22 In testing this, I found a few more issues. `policyTagList` can be an empty object, if there are no tags. I updated types to reflect that.

Similarly, right now we're assuming `transactionViolations` is always going to be defined, but there aren't very solid guarantees of that. We're relying on `policy` being set to mean that `transactionViolations` is also going to exist in onyx on updates and passing that directly through. I tried to see what a non-existent `transactionViolations` would return from `withOnyx` (i.e., is it `undefined` or an empty array or object), but I found that even on non-policy requests, they were getting `transactionViolations` keys with `cashExpenseWithNoReceipt`. Is that expected, or do we need to handle potentially empty `transactionViolations` values?

lindboe replied:
FWIW I would generally recommend making most types nullable in a case where we don't have full type-safety across the project yet, since you can't guarantee what's being passed in, but I'm not sure that's the approach being taken in the TS migration.

lindboe committed:
Feedback: Don't use Maybe type

lindboe committed:
Ensure an array is always passed to getViolationsOnyxData

lindboe committed:
Merge remote-tracking branch 'upstream/main' into lindboe/violations/money-request-updates

lindboe committed:
More prettier changes

lindboe committed:
Merge remote-tracking branch 'upstream/main' into lindboe/violations/money-request-updates

lindboe committed:
Merge remote-tracking branch 'upstream/main' into lindboe/violations/money-request-updates

lindboe commented:
@parasharrajat Main is merged 👍

lindboe commented on src/libs/actions/IOU.js:
```diff
@@ -1174,12 +1200,15 @@ function updateMoneyRequestMerchant(transactionID, transactionThreadReportID, va
+ * @param {Array} policyCategories
  */
-function updateMoneyRequestTag(transactionID, transactionThreadReportID, tag) {
+function updateMoneyRequestTag(transactionID, transactionThreadReportID, tag, policy, policyTags, policyCategories) {
```
Agh, yup, you're right, that was added in the refactor that just got merged in and so it was missed.

lindboe commented on src/libs/actions/IOU.js:
```diff
@@ -922,13 +922,16 @@ function createDistanceRequest(report, participant, comment, created, category,
  * @returns {object}
  */
-function getUpdateMoneyRequestParams(transactionID, transactionThreadReportID, transactionChanges, onlyIncludeChangedFields) {
+function getUpdateMoneyRequestParams(transactionID, transactionThreadReportID, transactionChanges, policy = {}, policyTags, policyCategories, onlyIncludeChangedFields) {
```
Agreed, I did do this originally, but the refactor added some I didn't know about. Will fix.

lindboe committed:
Merge remote-tracking branch 'upstream/main' into lindboe/violations/money-request-updates

lindboe committed:
Support client-side violations in new commands

lindboe commented:
@parasharrajat Update pushed, the newer commands should be included now.

> I noticed that I have to reselect either category or tag for outofPolicy violation to show up.

So the `outOfPolicy` violation showing up in Onyx is actually the responsibility of the backend, at this point. There's no code to respond to the policy configuration change being pushed, so we can only make client-side changes when the request itself is edited. (relevant slack thread here: https://expensify.slack.com/archives/C01GTK53T8Q/p1706134329225599?thread_ts=1706115595.439199&cid=C01GTK53T8Q). This means the client right now will only be responsible for removing the `outOfPolicy` violation in the optimistic data after a relevant update.

If you continue to see this issue, let us know so we can make Carlos aware of it, but be aware that there could potentially be some lag there.

lindboe commented:
@parasharrajat While I'm able to reproduce the "auto-set category" bug, I'm not able to reproduce the violation showing up for missing category anyways. I'm continuing to try. FWIW, I can trigger the bug by creating a request with a category, and then a request without a category -- the request without a category will receive the category from the last request I made. Seems like a bug with transaction drafts, going to check if it also exists on main.

lindboe commented:
@parasharrajat Looks like the auto-categorization is a server-side feature, so that's another issue this PR cannot address. https://expensify.slack.com/archives/C01GTK53T8Q/p1706221657329989, but I pinged you on the thread as well.

For issue 2, I would check in your Onyx database to check if policy updates are propagating to your client, (policyCategories_POLICYID/policyTags_POLICYID), but beyond that, I think we'll need @cead22 to help resolving with issues with policy data syncing.

cead22 commented:
@parasharrajat can you refresh and check if the violation goes away? I believe the reason it's not updating in there is because we're not sending violation data from UpdateMoneyRequestAmountAndCurrency which I'm fixing here https://github.com/Expensify/Expensify/issues/361948

lindboe commented:
> BUG: Looks like selecting the category does not remove the violation.

Okay yeah, I can reproduce that bug. This didn't happen before, so something new is wrong, looking into it

lindboe committed:
Pass policy args to all commands

lindboe commented:
@parasharrajat so sorry, it looks like I forgot to commit the update to the new commands where we pass in policy parameters in `EditRequestPage`. Pushed that update, and it should work as expected for you now:

https://github.com/Expensify/App/assets/5252268/e5ac0645-5549-4a4f-83ef-c38390c53849

parasharrajat reviewed (approved):
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.


:ribbon: :eyes: :ribbon:  C+ reviewed

cead22 commented:
@parasharrajat the videos don't show any of the tests, can you please update them to show the following?

> 1. Violations should be visible when viewing the request on each menu item with text and a RBR indicator. Here, we're generally concerned the following violations that we calculate client-side: missing category, missing tag, category out of policy, and tag out of policy. Other visible violations should not be affected by this PR.

cead22 commented:
> * If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack

I also noticed console warnings showing in the videos, can you report them on slack and link here please?

cead22 commented on src/libs/actions/IOU.js:
```diff
@@ -1079,6 +1082,14 @@ function getUpdateMoneyRequestParams(transactionID, transactionThreadReportID, t
     }
 
+    // Add optimistic transaction violations if there is a policy
+    const currentTransactionViolations = allTransactionViolations[`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`];
+    if (policy && policy.id) {
+        optimisticData.push(
+            ViolationsUtils.getViolationsOnyxData(updatedTransaction, currentTransactionViolations || [], policy.requiresTag, policyTags, policy.requiresCategory, policyCategories),
```
Can we move the fallback ` || []` to the line where we define `currentTransactionViolations`?
```suggestion
    const currentTransactionViolations = allTransactionViolations[`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`]  || [];
    if (policy && policy.id) {
        optimisticData.push(
            ViolationsUtils.getViolationsOnyxData(updatedTransaction, currentTransactionViolations, policy.requiresTag, policyTags, policy.requiresCategory, policyCategories),
```

cead22 commented on src/libs/actions/IOU.js:
```diff
@@ -2481,6 +2524,25 @@ function editRegularMoneyRequest(transactionID, transactionThreadReportID, trans
 
+    // Add transaction violations if there is a policy
+    if (policy && policy.id) {
+        const currentTransactionViolations = allTransactionViolations[`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`];
+        const updatedViolationsOnyxData = ViolationsUtils.getViolationsOnyxData(
+            updatedTransaction,
+            currentTransactionViolations || [],
```
Same here
```suggestion
        const currentTransactionViolations = allTransactionViolations[`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transactionID}`]  || [];
        const updatedViolationsOnyxData = ViolationsUtils.getViolationsOnyxData(
            updatedTransaction,
            currentTransactionViolations,
```

cead22 commented:
The onyx data changes reflect in the UI in the form of violations

lindboe committed:
Make currentTransactionViolations always an empty array if falsy

lindboe commented:
@parasharrajat At the time this PR was created, there were a number of issues with violations reliably showing up in the UI that were addressed in other PRs, so I tried to write tests to show this part of the feature still works. At this point @cead22 knows more than me about where those stand.

I see there's a conflict, I'll merge and re-test.

cead22 commented:
Violations are showing as you can see in some of the screenshots in the PR description. We should test that they're all showing on the right fields when applicable, and that they don't show when it doesn't apply, just like we're checking whether the onyx data is there or not

cead22 commented:
@parasharrajat I added dev.rajatparashar@gmail.com to the violation beta, but let me know if you want me to add any other test accounts

lindboe committed:
Merge remote-tracking branch 'upstream/main' into lindboe/violations/money-request-updates

lindboe commented:
@parasharrajat @cead22 I think this should be good to go now, conflict-wise

lindboe commented:
> Also, there are a couple of requested changes above. Please review them.

Are you referring to the request to move the `|| []` clause? that's been addressed.

lindboe committed:
Merge remote-tracking branch 'upstream/main' into lindboe/violations/money-request-updates

lindboe commented:
I've got a new issue now with `updateMoneyRequestBillable`, because the component which uses it, `MoneyRequestView`, is typed, but `actions/IOU` is not, and it's giving a type error due to bad inference:

<img width="1660" alt="Screenshot 2024-02-05 at 10 30 19 AM" src="https://github.com/Expensify/App/assets/5252268/3d2b1aa9-10a4-42bb-9bb9-ff04f84ff055">

I'm pushing a version with ts-ignore, since this is actually about incomplete types, but the other options are:

1. remove policy args from this action, since I don't think this will ever add client-side violations (which means we'd need to make sure `getUpdateMoneyRequestParams` behaves correctly with default arguments, which is fine)
2. Type the `IOU` file, which I expect would be a lot of work

@parasharrajat @cead22 let me know your thoughts.

lindboe committed:
Manage MoneyRequestView type error

This component was typed before the actions file, and so it's not
inferring the correct type for the policy arg. Should be fixed when
IOU.js is typed.

cead22 commented:
Nice, even better

lindboe committed:
Merge remote-tracking branch 'upstream/main' into lindboe/violations/money-request-updates

lindboe commented:
Looks like in addition to the type changes, more usages of these functions have come in with addition IOU refactors, like the addition of the `IOURequestStepTag` component: https://github.com/Expensify/App/pull/35136.

This is going to be a bigger merge with substantial changes.

lindboe committed:
Remove obsolete ts-expect-error

lindboe commented:
This is going to be a more substantial update. In addition to needing to update types differently now, there was also at least one refactor that moved the call site of a changed command (https://github.com/Expensify/App/pull/35136) and I'll have to fix those as well.

lindboe committed:
First stab at fixing typing

Problems remaining:
1. Inconsistent types
2. Mismatch between Onyx types and what's actually sent, and my code
   based on what's actually sent doesn't account for bad types:

```
> new.expensify@1.4.36-5 typecheck
> tsc

src/components/ReportActionItem/MoneyRequestView.tsx:159:109 - error TS2345: Argument of type 'OnyxEntry<Policy>' is not assignable to parameter of type 'Policy'.
  Type 'null' is not assignable to type 'Policy'.

159             IOU.updateMoneyRequestBillable(transaction?.transactionID ?? '', report?.reportID, newBillable, policy, policyTags, policyCategories);
                                                                                                                ~~~~~~

Found 1 error in src/components/ReportActionItem/MoneyRequestView.tsx:159
```

It's never actually null, it's empty object

3. Need to double-check that everything is set by default when it needs
   to be

lindboe committed:
WIP update

Partway through updating code, realizing that `updateMoneyRequest` is
now a completely dead code branch

lindboe committed:
editMoneyRequest is no longer used, delete

lindboe committed:
Finish typing IOU

lindboe commented:
I've tried to update with latest main, and I've run into two problems:

1. `policyTags` types not matching what's actually returned by Onyx (which was previously discussed [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1704929072385369) is now causing problems now that `MoneyRequestView` is typed; and I can't give it the correct type of `PolicyTagList` without updating the corresponding key in `src/ONYXKEYS.ts`. I'm either going to have to ignore the type errors in MoneyRequestView or re-do the typing in PolicyUtils (and where it reaches, like actions/Policy.ts and ModifiedExpenseMessage.ts. @cead22 do you have input on a preferred approach here?
2. `editMoneyRequest`, which I've been keeping up to date, now looks like a totally dead code path. I pushed a commit to remove it instead of continuing to update it, but I can re-add it if needed.

@parasharrajat @cead22 please let me know if you have opinions on either of these items

lindboe commented:
> Eventually, we will migrate away to standalone update functions from this.

I think that's already happened, that's why `editMoneyRequest` is no longer called by anything (and why I deleted it).

lindboe committed:
Merge remote-tracking branch 'upstream/main' into lindboe/violations/money-request-updates

lindboe committed:
Expect error in MoneyRequestView

lindboe committed:
Random lint change

lindboe committed:
Revert "editMoneyRequest is no longer used, delete"

This reverts commit 34930401a0b74bc3144a000b05fb3da620d2a865.

lindboe commented:
Okay, so that this can move forward, I'm aiming for minimal changes:

1. I un-deleted EditMoneyRequest since there's a deletion test now using it
2. I added a `@ts-expect-error` to the type issue in `MoneyRequestView` with the conflict between the component's typing as `PolicyTags` (not correct type for what the data returns and the ViolationsUtils function's use of `PolicyTagList`). I'll make a more clear report of this and share it in open-source so someone can work on fixing the policy tags typing and utils.
3. I changed the policy* defaultProps to `null` where TypeScript was complaining because `OnyxEntry<Policy*> = Policy* | null`, but defaults props was returning an empty object, and this difference was leading to some very confusing type conflicts. Based on my conversation [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1707194455003059?thread_ts=1707175571.145789&cid=C01GTK53T8Q), `null` is the expected "empty" data type once components are typescript, so may as well change it now instead of trying to handle both everywhere.

Going to run through some test cases and make sure it's all good now.

lindboe commented:
Okay @parasharrajat this should be good to go again, except the perf-tests that appear to be failing on main as well

lindboe commented:
Asked about policyTags types here: https://expensify.slack.com/archives/C01GTK53T8Q/p1707263531033829

cead22 commented:
> 1. @cead22 do you have input on a preferred approach here?

Thanks, I was going to suggest what you already did (to ignore the type errors for now, since the more we add to this PR the harder it'll get to get it merged)

lindboe committed:
Merge remote-tracking branch 'upstream/main' into lindboe/violations/money-request-updates

lindboe committed:
Update command in new description component

lindboe commented:
Another day, another refactor update: the command to update description now lives in its own component. Updated.

@cead22 while testing this update, I noticed an assumption in earlier decisions doesn't seem to be true: in a few places, we're checking if `policy?.id` exists, because "if it's a P2P request, there won't be a policy". That may have been true at one point, but isn't now, I do see a policy when I make a P2P request:

```
{
    "isFromFullPolicy": false,
    "id": "B88059F8975D1B23",
    "name": "Lizzi's Expenses",
    "role": "...",
    "type": "personal",
    "owner": "...",
    "ownerAccountID": ...,
    "outputCurrency": "USD",
    "avatar": "",
    "employeeList": [],
    "isPolicyExpenseChatEnabled": false,
    "lastModified": ...,
    "chatReportIDAnnounce": 0,
    "chatReportIDAdmins": 0
}
```

I'm assuming `isFromFullPolicy` or `type` are fields we should be checking against instead?

lindboe commented:
Above issue with `policy?.id` discussed and resolved here: https://expensify.slack.com/archives/C01GTK53T8Q/p1707339619099769.

@parasharrajat tests are passing again. Just to clarify, next steps were for you to test this out again right?

parasharrajat reviewed (approved):
LGTM. There are couple of prop warnings but I think those will be removed when these components are migrated to TS.

I will also report some to Slack in sometime.

lindboe committed:
Merge remote-tracking branch 'upstream/main' into lindboe/violations/money-request-updates

cead22 commented on src/libs/Violations/ViolationsUtils.ts:
```diff
@@ -24,15 +24,16 @@ const ViolationsUtils = {
             const hasMissingCategoryViolation = transactionViolations.some((violation) => violation.name === 'missingCategory');
-            const isCategoryInPolicy = !!policyCategories[transaction.category ?? '']?.enabled;
+            const categoryKey = updatedTransaction.category;
+            const isCategoryInPolicy = categoryKey ? policyCategories?.[categoryKey]?.enabled : undefined;
```
Can/should we fallback to `false` here instead of `undefined`? I think false is a better value for `isCategoryInPolicy`
```suggestion
            const isCategoryInPolicy = categoryKey ? policyCategories?.[categoryKey]?.enabled : false;
```

lindboe replied:
Sure, that's fine.  I think it can be a fine semantic distinction between "not in policy' and "policy category missing", but this will work too.

cead22 commented on src/libs/Violations/ViolationsUtils.ts:
```diff
@@ -42,23 +43,25 @@ const ViolationsUtils = {
             const hasTagOutOfPolicyViolation = transactionViolations.some((violation) => violation.name === 'tagOutOfPolicy');
             const hasMissingTagViolation = transactionViolations.some((violation) => violation.name === 'missingTag');
-            const isTagInPolicy = !!policyTags[transaction.tag ?? '']?.enabled;
+            const isTagInPolicy = policyTags ? !!policyTags[updatedTransaction.tag ?? '']?.enabled : undefined;
```
Same here
```suggestion
            const isTagInPolicy = policyTags ? !!policyTags[updatedTransaction.tag ?? '']?.enabled : false;
```

lindboe committed:
Apply suggestions from code review

Co-authored-by: Carlos Alvarez <cead22@gmail.com>

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7872745855) to staging by https://github.com/cead22 in version: 1.4.40-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|failure ❌

mountiny commented:
Noting here that QA did not really validate this PR as the QA steps were not clear enough for them it seems like. I believe its fine not to block deploy on this and we will internally test these and report any issues as we find them

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7890953833) to production by https://github.com/mountiny in version: 1.4.40-5 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|failure ❌
